src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,461 @@
+
+        if (what == "vow") vow = data;
+        else if (what == "end") end = EndLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "tau" ) {
+            ilks[ilk].tau = data;
+        } else revert("DssDirectDepositHub/file-unrecognized-param");
+
+        emit File(ilk, what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "pool") ilks[ilk].pool = D3MPoolLike(data);
+        else if (what == "plan") ilks[ilk].plan = D3MPlanLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, D3MPoolLike pool, uint256 amount) internal {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+        if (amount == 0) {
+            emit Wind(ilk, 0);
+            return;
+        }
+
+        require(int256(amount) >= 0, "DssDirectDepositHub/overflow");
+
+        vat.slip(ilk, address(pool), int256(amount));
+        vat.frob(ilk, address(pool), address(pool), address(this), int256(amount), int256(amount));
+        // normalized debt == erc20 DAI (Vat rate for this ilk fixed to 1 RAY)
+        daiJoin.exit(address(pool), amount);
+        pool.deposit(amount);
+
+        emit Wind(ilk, amount);
+    }
+
+    function _unwind(bytes32 ilk, D3MPoolLike pool, uint256 supplyReduction, uint256 availableAssets, Mode mode, uint256 assetBalance) internal {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why it converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        EndLike end_;
+        uint256 daiDebt;
+        if (mode == Mode.NORMAL) {
+            // Normal mode or module just caged (no culled)
+            // debt is obtained from CDP art
+            (,daiDebt) = vat.urns(ilk, address(pool));
+        } else if (mode == Mode.MODULE_CULLED) {
+            // Module shutdown and culled
+            // debt is obtained from free collateral owned by this contract
+            daiDebt = vat.gem(ilk, address(pool));
```

Comment:
So there is a bit of semantical error here and in the else case.
Marking the `daiDebt` as the amount of gems is wrong as it could have happened that someone repaid debt to this position (as it is permissionless) during normal operations.
That provokes the actual `daiDebt` decreases, meaning that when is compared to the `assetbalance` there are more fees to send to the surplus.
Basically from that moment there will be more `gem`s in the vat than actual `dai` deposited. Well there might be more if fees are accrued but will balance out to a lower amount.
If that scenario happens and then shutdown occurs, `daiDebt` will end up with a value higher than the `assetBalance`. In pratice it is not an issue as `amount` will be restricted by `assetBalance` anyway but I think from a semantically point of view we should not assign the gem balance to `daiDebt`, but just assign it as `type(uint256).max` or maybe the `assetBalance` itself.
There is not a bug due the `assetBalance` `min` ceiling, and it comes from the old code, but I think we should make this thing cleaner now.

Reply:
@gbalabasquer based on our conversation this week I made these changes to add rebalancing to cull so we are correctly addressing daiDebt in the culled unwind: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/dc0ae0d06e594527dae14c4cc8b83e823d7761ce

Reply:
Looks good!